keps/sig-network/1864-disable-lb-node-ports/README.md

```diff
@@ -0,0 +1,153 @@
+  - if `allocateLoadBalancerNodePort: false`, and a user wants to disable node ports on existing Services, they need to remove the nodePort field explicitly.
+
+When a user relies on `kubectl apply` to fill in the node port field, the node port would never be deallocated since the existing node ports from the
+server will always be merged prior to update. A user must send an explicit update request for the Service using something like `kubectl edit svc` or
+building a controller to update all Services in a cluster.
+
+### Risks and Mitigations
+
+A user may unknowingly disable node ports while it is serving traffic for their pods. The chances of this should be significantly reduced since
+node ports are not automatically deallocated when a user sets `allocateLoadBalancerNodePort: false`. The additional step to disable node ports
+should ensure users are aware of the consequences of this change.
+
+## Design Details
+
+API changes to Service:
+* Add a new field `spec.allocateLoadBalancerNodePorts: true|false`.
+* `allocateLoadBalancerNodePorts` defaults to true, preserving existing behavior for Service Type=LoadBalancer.
+* On create, if `allocateLoadBalancerNodePorts: false`, don't allocate node ports.
+* On update, if `allocateLoadBalancerNodePorts: false` don't allocate new node ports but do not deallocate existing node ports if set.
+* On update, set `nodePort: 0` (by removing the field) to deallocate a node port, a new port will not be re-allocated if `allocateLoadBalancerNodePorts: false`.
+* On delete, node ports are deallocated regardless of `allocateLoadBalancerNodePorts`.
+
+### Test Plan
+
+Unit tests:
+* unit tests for the allocate/deallocation of node ports based on the value of `allocateLoadBalancerNodePorts`.
+* validate the default value for `allocateLoadBalanceNodePorts` on Service.
+
+E2E tests:
+* The default behavior for `allocateLoadBalancerNodePorts` does not break any existing e2e tests.
+* e2e tests for disabling node ports on a Service (testing the network data path may be difficult since testing platforms uses node ports).
+* e2e tests for disabling node ports on an existing Service, ensure the node ports are preserved.
+* e2e tests to explicitly disable node ports with `allocateLoadBalancerNodePorts: false`.
+* e2e tests to re-enable `allocateLoadBalancerNodeports` for a Service with node ports disabled.
+
+### Graduation Criteria
+
+### GA
+
+This feature is for optionally disabling an existing feature so it does not require alpha/beta stages of graduation.
+
+### Upgrade / Downgrade Strategy
+
+Upgrade should be trivial since kube-proxy's behavior of the Service is determined by whether a node port is set.
+For existing Services that are upgraded, `nodePort` will continue to be set (even if `allocateLoadBalancerNodePort: false`).
+
+On downgrade, if `allocateLoadBalancerNodePorts: false`, the worse case is that a Service which was intended to have
```

Comment:
I'm worried that the worst case is that something which assumes that NodePorts MUST be set will explode  when they are not.  It's like changing an invariant - someone is relying on that assumption. :(